tool/src/cpp/docs.rs

```diff
@@ -0,0 +1,232 @@
+        {
+            let out = outs
+                .entry(format!("{}.rst", in_path.elements.join("_")))
+                .or_insert_with(String::new);
+
+            let title = format!("``{}``", in_path.elements.join("::"));
+            writeln!(out, "{}", title)?;
+            writeln!(out, "{}", "=".repeat(title.len()))?;
+            writeln!(out)?;
+
+            let mut sorted_symbols: Vec<&String> = mod_symbols.keys().collect();
+            sorted_symbols.sort();
+
+            for symbol_name in sorted_symbols {
+                let custom_type = &mod_symbols[symbol_name];
+                if let ast::ModSymbol::CustomType(typ) = custom_type {
+                    writeln!(out)?;
+                    gen_custom_type_docs(out, typ, in_path, env)?;
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+pub fn gen_custom_type_docs<W: fmt::Write>(
+    out: &mut W,
+    typ: &ast::CustomType,
+    in_path: &ast::Path,
+    env: &HashMap<ast::Path, HashMap<String, ast::ModSymbol>>,
+) -> fmt::Result {
+    match typ {
+        ast::CustomType::Struct(_) => writeln!(out, ".. cpp:struct:: {}", typ.name())?,
+        ast::CustomType::Enum(_) => writeln!(out, ".. cpp:enum-struct:: {}", typ.name())?,
+        ast::CustomType::Opaque(_) => writeln!(out, ".. cpp:class:: {}", typ.name())?,
+    }
+
+    writeln!(out)?;
+    let mut class_indented = indented(out).with_str("    ");
+    markdown_to_rst(
+        &mut class_indented,
+        typ.doc_lines(),
+        &|shortcut_path, to| {
+            let resolved = ast::TypeName::Named(shortcut_path.clone()).resolve(in_path, env);
+            match resolved {
+                ast::CustomType::Struct(_) => write!(to, ":cpp:struct:`{}`", resolved.name())?,
+                ast::CustomType::Enum(_) => write!(to, ":cpp:enum-struct:`{}`", resolved.name())?,
+                ast::CustomType::Opaque(_) => write!(to, ":cpp:class:`{}`", resolved.name())?,
+            }
+            Ok(())
+        },
+    )?;
+    writeln!(class_indented)?;
+
+    if let ast::CustomType::Struct(strct) = typ {
+        for field in strct.fields.iter() {
+            writeln!(&mut class_indented)?;
+            gen_field_docs(&mut class_indented, field, in_path, env)?;
+        }
+    } else if let ast::CustomType::Enum(enm) = typ {
+        for variant in &enm.variants {
+            writeln!(&mut class_indented)?;
+            gen_enum_variant_docs(&mut class_indented, variant, in_path, env)?;
+        }
+    }
+
+    for method in typ.methods() {
+        writeln!(&mut class_indented)?;
+        gen_method_docs(&mut class_indented, method, in_path, env)?;
+    }
+    Ok(())
+}
+
+pub fn gen_method_docs<W: fmt::Write>(
+    out: &mut W,
+    method: &ast::Method,
+    in_path: &ast::Path,
+    env: &HashMap<ast::Path, HashMap<String, ast::ModSymbol>>,
+) -> fmt::Result {
+    let mut params: Vec<Param> = method.params.clone();
+    if method.is_writeable_out() {
```

Comment:
Probably should handle the templated Writeable out as well?

Reply:
Also, shouldn't this change the return type as well?

Reply:
fixed!